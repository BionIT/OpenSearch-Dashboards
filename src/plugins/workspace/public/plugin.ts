/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  createOsdUrlStateStorage,
  createStateContainer,
  syncState,
} from '../../opensearch_dashboards_utils/public';
import { CoreStart, Plugin } from '../../../core/public';

export class WorkspacePlugin implements Plugin<{}, {}, {}> {
  public async setup() {
    return {};
  }

  public start(core: CoreStart) {
    const initialState = {
      id: 'foo',
    };
    const globalQueryStateContainer = createStateContainer({});
    const urlStateStorageChange = createOsdUrlStateStorage({
      history: core.application.history,
      useHash: false,
    });
    const { start } = syncState({
      /**
       * Hijack urlStateStorage to call set with replace: true
       */
      stateStorage: {
        ...urlStateStorageChange,
        set(key, state) {
          return urlStateStorageChange.set(key, state, {
            replace: true,
          });
        },
      },
      stateContainer: {
        ...globalQueryStateContainer,
        set: (state) => {
          globalQueryStateContainer.set({
            ...initialState,
            ...state,
          });
        },
        get: () => {
          /**
           * Adding a timestamp field to make each response non-idempotent.
           * Or the syncState will compare each state, if no change it won't trigger stateContainer.set
           */
          return {
            ...initialState,
            timestamp: Date.now(),
          };
        },
      },
      storageKey: '_w',
    });
    start();
    return {};
  }

  public stop() {}
}
