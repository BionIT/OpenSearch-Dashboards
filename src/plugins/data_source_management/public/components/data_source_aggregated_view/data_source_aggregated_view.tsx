/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import {
  EuiButtonEmpty,
  EuiButtonIcon,
  EuiContextMenu,
  EuiNotificationBadge,
  EuiPopover,
} from '@elastic/eui';
import { i18n } from '@osd/i18n';
import { SavedObjectsClientContract, ToastsStart } from 'opensearch-dashboards/public';
import { DataSourceOption } from '../data_source_selector/data_source_selector';
import { getDataSourcesWithFields } from '../utils';

interface DataSourceAggregatedViewProps {
  savedObjectsClient: SavedObjectsClientContract;
  notifications: ToastsStart;
  hideLocalCluster: boolean;
  fullWidth: boolean;
  activeDatasourceIds?: string[];
  filterFn?: (dataSource: any) => boolean;
  displayAllCompatibleDataSources: boolean;
}

interface DataSourceAggregatedViewState {
  dataSourceOptions: DataSourceOption[];
  selectedOption: DataSourceOption[];
  isPopoverOpen: boolean;
  allDataSourcesIdMap: Map<string, any>;
}

export class DataSourceAggregatedView extends React.Component<
  DataSourceAggregatedViewProps,
  DataSourceAggregatedViewState
> {
  private _isMounted: boolean = false;

  constructor(props: DataSourceAggregatedViewProps) {
    super(props);

    this.state = {
      isPopoverOpen: false,
      allDataSourcesIdMap: new Map(),
    };
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  onClick() {
    this.setState({ ...this.state, isPopoverOpen: !this.state.isPopoverOpen });
  }

  closePopover() {
    this.setState({ ...this.state, isPopoverOpen: false });
  }

  async componentDidMount() {
    this._isMounted = true;
    getDataSourcesWithFields(this.props.savedObjectsClient, ['id', 'title', 'auth.type'])
      .then((fetchedDataSources) => {
        if (fetchedDataSources?.length) {
          let filteredDataSources = fetchedDataSources;
          if (this.props.filterFn) {
            filteredDataSources = fetchedDataSources.filter((ds) => this.props.filterFn!(ds));
          }

          const allDataSourcesIdMap = new Map();

          filteredDataSources.forEach((ds) => {
            allDataSourcesIdMap.set(ds.id, ds.attributes!.title || '');
          });

          if (!this.props.hideLocalCluster) {
            allDataSourcesIdMap.set('', 'Local cluster');
          }

          if (!this._isMounted) return;
          this.setState({
            ...this.state,
            allDataSourcesIdMap,
          });
        }
      })
      .catch(() => {
        this.props.notifications.addWarning(
          i18n.translate('dataSource.fetchDataSourceError', {
            defaultMessage: 'Unable to fetch existing data sources',
          })
        );
      });
  }

  render() {
    const button = (
      <EuiButtonIcon
        iconType="iInCircle"
        display="empty"
        aria-label="Next"
        onClick={this.onClick.bind(this)}
      />
    );

    let items = [];

    // only display active data sources
    if (this.props.activeDatasourceIds && this.props.activeDatasourceIds.length > 0) {
      items = this.props.activeDatasourceIds.map((id) => {
        return {
          name: this.state.allDataSourcesIdMap.get(id),
          disabled: true,
        };
      });
    } else {
      items = [...this.state.allDataSourcesIdMap.values()].map((v) => {
        return {
          name: v,
          disabled: true,
        };
      });
    }

    const title = this.props.displayAllCompatibleDataSources
      ? `Data sources (${this.state.allDataSourcesIdMap.size})`
      : 'Selected data sources';

    const panels = [
      {
        id: 0,
        title,
        items,
      },
    ];

    return (
      <>
        <EuiButtonEmpty
          className="euiHeaderLink"
          data-test-subj="dataSourceAggregatedViewContextMenuHeaderLink"
          aria-label={i18n.translate('dataSourceAggregatedView.dataSourceOptionsButtonAriaLabel', {
            defaultMessage: 'dataSourceAggregatedViewMenuButton',
          })}
          iconType="database"
          iconSide="left"
          size="s"
          disabled={true}
        >
          {'Data sources'}
        </EuiButtonEmpty>
        <EuiNotificationBadge color={'subdued'}>
          {this.props.activeDatasourceIds?.length || 'All'}
        </EuiNotificationBadge>
        <EuiPopover
          id={'dataSourceSViewContextMenuPopover'}
          button={button}
          isOpen={this.state.isPopoverOpen}
          closePopover={this.closePopover.bind(this)}
          panelPaddingSize="none"
          anchorPosition="downLeft"
        >
          <EuiContextMenu initialPanelId={0} panels={panels} />
        </EuiPopover>
      </>
    );
  }
}
