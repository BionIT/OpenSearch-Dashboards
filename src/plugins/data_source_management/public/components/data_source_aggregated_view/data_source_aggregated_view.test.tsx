/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { ShallowWrapper, shallow } from 'enzyme';
import React from 'react';
import { DataSourceAggregatedView } from './data_source_aggregated_view';
import { SavedObjectsClientContract } from '../../../../../core/public';
import { notificationServiceMock } from '../../../../../core/public/mocks';
import { getDataSourcesWithFieldsResponse, mockResponseForSavedObjectsCalls } from '../../mocks';

describe('DataSourceAggregatedView', () => {
  let component: ShallowWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;
  let client: SavedObjectsClientContract;
  const { toasts } = notificationServiceMock.createStartContract();

  beforeEach(() => {
    client = {
      find: jest.fn().mockResolvedValue([]),
    } as any;
    mockResponseForSavedObjectsCalls(client, 'find', getDataSourcesWithFieldsResponse);
  });

  it('should render normally with local cluster not hidden and all options', () => {
    component = shallow(
      <DataSourceAggregatedView
        fullWidth={false}
        hideLocalCluster={false}
        savedObjectsClient={client}
        notifications={toasts}
        displayAllCompatibleDataSources={true}
      />
    );
    expect(component).toMatchSnapshot();
    expect(client.find).toBeCalledWith({
      fields: ['id', 'title', 'auth.type'],
      perPage: 10000,
      type: 'data-source',
    });
    expect(toasts.addWarning).toBeCalledTimes(0);
  });

  it('should render normally with local cluster hidden and all options', () => {
    component = shallow(
      <DataSourceAggregatedView
        fullWidth={false}
        hideLocalCluster={true}
        savedObjectsClient={client}
        notifications={toasts}
        displayAllCompatibleDataSources={true}
      />
    );
    expect(component).toMatchSnapshot();
    expect(client.find).toBeCalledWith({
      fields: ['id', 'title', 'auth.type'],
      perPage: 10000,
      type: 'data-source',
    });
    expect(toasts.addWarning).toBeCalledTimes(0);
  });

  it('should render normally with local cluster and actice selections', () => {
    component = shallow(
      <DataSourceAggregatedView
        fullWidth={false}
        hideLocalCluster={false}
        savedObjectsClient={client}
        notifications={toasts}
        displayAllCompatibleDataSources={false}
        activeDatasourceIds={['test1']}
      />
    );
    expect(component).toMatchSnapshot();
    expect(client.find).toBeCalledWith({
      fields: ['id', 'title', 'auth.type'],
      perPage: 10000,
      type: 'data-source',
    });
    expect(toasts.addWarning).toBeCalledTimes(0);
  });
});
