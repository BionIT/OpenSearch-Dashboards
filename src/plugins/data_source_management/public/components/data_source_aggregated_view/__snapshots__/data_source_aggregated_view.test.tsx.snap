// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataSourceAggregatedView should render normally with local cluster and actice selections 1`] = `
<Fragment>
  <EuiButtonEmpty
    aria-label="dataSourceAggregatedViewMenuButton"
    className="euiHeaderLink"
    data-test-subj="dataSourceAggregatedViewContextMenuHeaderLink"
    disabled={true}
    iconSide="left"
    iconType="database"
    size="s"
  >
    Data sources
  </EuiButtonEmpty>
  <EuiNotificationBadge
    color="subdued"
  >
    1
  </EuiNotificationBadge>
  <EuiPopover
    anchorPosition="downLeft"
    button={
      <EuiButtonIcon
        aria-label="Next"
        display="empty"
        iconType="iInCircle"
        onClick={[Function]}
      />
    }
    closePopover={[Function]}
    display="inlineBlock"
    hasArrow={true}
    id="dataSourceSViewContextMenuPopover"
    isOpen={false}
    ownFocus={true}
    panelPaddingSize="none"
  >
    <EuiContextMenu
      initialPanelId={0}
      panels={
        Array [
          Object {
            "id": 0,
            "items": Array [
              Object {
                "disabled": true,
                "name": undefined,
              },
            ],
            "title": "Selected data sources",
          },
        ]
      }
      size="m"
    />
  </EuiPopover>
</Fragment>
`;

exports[`DataSourceAggregatedView should render normally with local cluster hidden and all options 1`] = `
<Fragment>
  <EuiButtonEmpty
    aria-label="dataSourceAggregatedViewMenuButton"
    className="euiHeaderLink"
    data-test-subj="dataSourceAggregatedViewContextMenuHeaderLink"
    disabled={true}
    iconSide="left"
    iconType="database"
    size="s"
  >
    Data sources
  </EuiButtonEmpty>
  <EuiNotificationBadge
    color="subdued"
  >
    All
  </EuiNotificationBadge>
  <EuiPopover
    anchorPosition="downLeft"
    button={
      <EuiButtonIcon
        aria-label="Next"
        display="empty"
        iconType="iInCircle"
        onClick={[Function]}
      />
    }
    closePopover={[Function]}
    display="inlineBlock"
    hasArrow={true}
    id="dataSourceSViewContextMenuPopover"
    isOpen={false}
    ownFocus={true}
    panelPaddingSize="none"
  >
    <EuiContextMenu
      initialPanelId={0}
      panels={
        Array [
          Object {
            "id": 0,
            "items": Array [],
            "title": "Data sources (0)",
          },
        ]
      }
      size="m"
    />
  </EuiPopover>
</Fragment>
`;

exports[`DataSourceAggregatedView should render normally with local cluster not hidden and all options 1`] = `
<Fragment>
  <EuiButtonEmpty
    aria-label="dataSourceAggregatedViewMenuButton"
    className="euiHeaderLink"
    data-test-subj="dataSourceAggregatedViewContextMenuHeaderLink"
    disabled={true}
    iconSide="left"
    iconType="database"
    size="s"
  >
    Data sources
  </EuiButtonEmpty>
  <EuiNotificationBadge
    color="subdued"
  >
    All
  </EuiNotificationBadge>
  <EuiPopover
    anchorPosition="downLeft"
    button={
      <EuiButtonIcon
        aria-label="Next"
        display="empty"
        iconType="iInCircle"
        onClick={[Function]}
      />
    }
    closePopover={[Function]}
    display="inlineBlock"
    hasArrow={true}
    id="dataSourceSViewContextMenuPopover"
    isOpen={false}
    ownFocus={true}
    panelPaddingSize="none"
  >
    <EuiContextMenu
      initialPanelId={0}
      panels={
        Array [
          Object {
            "id": 0,
            "items": Array [],
            "title": "Data sources (0)",
          },
        ]
      }
      size="m"
    />
  </EuiPopover>
</Fragment>
`;
